dhcpv6/option_vendor_opts.go

```diff
@@ -0,0 +1,161 @@
+type OptVendorOpts struct {
+	EnterpriseNumber uint32
+	VendorOpts       []Option
+}
+
+// Code returns the option code
+func (op *OptVendorOpts) Code() OptionCode {
+	return OptionVendorOpts
+}
+
+// ToBytes serializes the option and returns it as a sequence of bytes
+func (op *OptVendorOpts) ToBytes() []byte {
+	buf := make([]byte, 8)
+	binary.BigEndian.PutUint16(buf[0:2], uint16(OptionVendorOpts))
+	binary.BigEndian.PutUint16(buf[2:4], uint16(op.Length()))
+	binary.BigEndian.PutUint32(buf[4:8], uint32(op.EnterpriseNumber))
+	for _, opt := range op.VendorOpts {
+		buf = append(buf, opt.ToBytes()...)
+	}
+	return buf
+}
+
+// Length returns the option length
+func (op *OptVendorOpts) Length() int {
+	l := 4 // 4 bytes for Enterprise Number
+	for _, opt := range op.VendorOpts {
+		l += 4 + opt.Length() // 4 bytes for Code and Length from Vendor
+	}
+	return l
+}
+
+// String returns a string representation of the VendorOpts data
+func (op *OptVendorOpts) String() string {
+	return fmt.Sprintf("OptVendorOpts{enterprisenum=%v, vendorOpts=%v}",
+		op.EnterpriseNumber, op.VendorOpts,
+	)
+}
+
+// ParseOptVendorOpts builds an OptVendorOpts structure from a sequence of bytes.
+// The input data does not include option code and length bytes.
+func ParseOptVendorOpts(data []byte) (*OptVendorOpts, error) {
+	opt := OptVendorOpts{}
+	if len(data) < 4 {
+		return nil, fmt.Errorf("Invalid vendor opts data length. Expected at least 4 bytes, got %v", len(data))
+	}
+	opt.EnterpriseNumber = binary.BigEndian.Uint32(data[:4])
+
+	var err error
+	opt.VendorOpts, err = VendorOptionsFromBytes(data[4:])
+	if err != nil {
+		return nil, err
+	}
+	return &opt, nil
+}
+
+// VendorOptionsFromBytes builds a slice of GenericOptions from a slice of bytes
+func VendorOptionsFromBytes(data []byte) ([]Option, error) {
+	// Parse a sequence of bytes until the end and build a list of options from
+	// it. Returns an error if any invalid option or length is found.
+	options := make([]Option, 0, 10)
+	if len(data) == 0 {
+		// no options, no party
+		return options, nil
+	}
+	if len(data) < 4 {
+		// cannot be shorter than option code (2 bytes) + length (2 bytes)
+		return nil, fmt.Errorf("Invalid options: shorter than 4 bytes")
+	}
+	idx := 0
+	for {
+		if idx == len(data) {
+			break
+		}
+		if idx > len(data) {
+			// this should never happen
+			return nil, fmt.Errorf("Error: reading past the end of options")
+		}
+		opt, err := VendParseOption(data[idx:])
+		if err != nil {
+			return nil, err
+		}
+		options = append(options, opt)
+		idx += opt.Length() + 4 // 4 bytes for type + length
+	}
+	return options, nil
+}
+
+// VendParseOption builds a GenericOption from a slice of bytes
+// We cannot use the exisitng ParseOption function in options.go because the
+// sub-options include codes specific to each vendor. There are overlaps in these
+// codes with RFC standard codes.
+func VendParseOption(dataStart []byte) (Option, error) {
```

Comment:
Isn't this the same as `OptionsFromBytes`?

https://github.com/insomniacslk/dhcp/blob/master/dhcpv6/options.go#L122

Reply:
Almost. Vendor options don't follow the iana codes, so for example we can have a vendor option number 1 which doesn't follow the structure of the Client ID option. That is why we can't use `OptionsFromBytes`.

Reply:
So the difference would be only that `OptionsFromBytes` calls internally `ParseOption`, but the latter is not what we need?
If so, we can generalize it as follows:
* `OptionsFromBytes` is split into `OptionsFromBytesWithParser` (or better name) that takes a parser function,. and the original `OptionsFromBytes` just calls the new function passing `ParseOption` as function pointer
* Instead of calling `VendorParseOption`, it will call `OptionsFromBytesWithParser` with a custom anonymous function that simply assigns `opt.OptionCode` and `opt.OptionData`.
How does that sound?

Reply:
Sounds good. We will need to define a new type for the parser function, won't we?

Reply:
Yes. Even if it sounds overcomplicated, I think it's still better than duplicating and testing that code again. We can see how it comes and decide

Reply:
To add some light into it, if I am not mistaken, I think we should define something like:
```
type Parser func(dataStart []byte) (Option, error)
```